llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp

```diff
@@ -7207,6 +7207,16 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
     setValue(&I, Res);
     return;
   }
+  case Intrinsic::cheri_cap_from_pointer: {
+    Value *Offset = I.getArgOperand(1);
+    // (int_cheri_cap_from_pointer x, 0) -> null
```

Comment:
SelectionDAGBuilder::visitIntrinsicCall is for lowering/legalisation, not optimisation? The cheri_cap_address_set code above is only run if the target doesn't support the intrinsic (which we should GC as an option at this point...).

Reply:
That's a good point - I only added it there since I didn't see intrinsics being handled anywhere else. Do you have a suggestion which function I should add this to?

Reply:
INTRINSIC_WO_CHAIN normally gets combined via FooTargetLowering::PerformDAGCombine. SelectionDAG just doesn't really have a good model for intrinsics given they're all mapped to the same ISD. I'd probably suggest just doing this in InstCombine given I doubt lowering introduces new instances of this pattern?

Reply:
Agreed, InstSimplify probably makes the most sense considering we currently won't create these in SelectionDAG.